File: app/pavel/utils/critical-path.ts
import type { Record, Link } from "../types"

interface ModelNode {
  id: string
  duration: number
  earliestStart: number
  earliestFinish: number
  latestStart: number
  latestFinish: number
  parents: string[]
  children: string[]
}

interface Graph {
  [key: string]: ModelNode
}

interface CalculationState {
  stage: string
  progress: number
}

export async function calculateCriticalPath(
  records: Record[],
  links: Link[],
  onProgress: (stage: string, progress: number) => void,
  signal?: AbortSignal,
): Promise<string[]> {
  console.log(`Calculating critical path for ${records.length} records and ${links.length} links`)

  function checkCancellation() {
    if (signal?.aborted) {
      throw new Error("Operation cancelled")
    }
  }

  try {
    // Create a Map of model durations for quick lookup
    const modelDurations = new Map(
      records.map((record) => [
        record.model,
        (new Date(record.completed_at).getTime() - new Date(record.started_at).getTime()) / 1000,
      ]),
    )

    // Build graph - O(V)
    onProgress("Building graph", 0)
    const graph: Graph = {}
    records.forEach((record) => {
      graph[record.model] = {
        id: record.model,
        duration: modelDurations.get(record.model) || 0,
        earliestStart: 0,
        earliestFinish: 0,
        latestStart: 0,
        latestFinish: 0,
        parents: [],
        children: [],
      }
    })

    // Process dependencies - O(E)
    onProgress("Processing dependencies", 25)
    links.forEach((link) => {
      if (graph[link.source] && graph[link.target]) {
        graph[link.source].children.push(link.target)
        graph[link.target].parents.push(link.source)
      }
    })

    // Kahn's algorithm for topological sort with earliest times calculation - O(V + E)
    onProgress("Calculating times", 50)
    const sorted: string[] = []
    const noIncoming = Object.keys(graph).filter((node) => graph[node].parents.length === 0)
    const inDegree = new Map(Object.keys(graph).map((node) => [node, graph[node].parents.length]))

    // Calculate earliest times during topological sort
    while (noIncoming.length > 0) {
      checkCancellation()
      const current = noIncoming.shift()!
      sorted.push(current)

      const node = graph[current]
      node.earliestFinish = node.earliestStart + node.duration

      for (const childId of node.children) {
        const child = graph[childId]
        child.earliestStart = Math.max(child.earliestStart, node.earliestFinish)

        const newDegree = inDegree.get(childId)! - 1
        inDegree.set(childId, newDegree)
        if (newDegree === 0) {
          noIncoming.push(childId)
        }
      }
    }

    // Calculate latest times in reverse order - O(V + E)
    onProgress("Finalizing critical path", 75)
    const projectEnd = Math.max(...Object.values(graph).map((node) => node.earliestFinish))

    // Initialize latest times
    sorted.reverse().forEach((nodeId) => {
      const node = graph[nodeId]
      if (node.children.length === 0) {
        node.latestFinish = projectEnd
      } else {
        node.latestFinish = Math.min(...node.children.map((childId) => graph[childId].latestStart))
      }
      node.latestStart = node.latestFinish - node.duration
    })

    // Find critical path (nodes where earliest = latest) - O(V)
    const criticalPath: string[] = []
    let current = sorted.find(
      (nodeId) =>
        graph[nodeId].children.length === 0 &&
        Math.abs(graph[nodeId].earliestFinish - graph[nodeId].latestFinish) < 0.001,
    )

    while (current) {
      checkCancellation()
      criticalPath.unshift(current)

      // Find the critical parent (if any)
      current = graph[current].parents.find(
        (parentId) =>
          Math.abs(graph[parentId].earliestFinish - graph[parentId].latestFinish) < 0.001 &&
          Math.abs(graph[parentId].earliestFinish - graph[current!].earliestStart) < 0.001,
      )
    }

    onProgress("Finalizing critical path", 100)
    console.log(`Critical path calculation complete. Path length: ${criticalPath.length}`)
    return criticalPath
  } catch (error) {
    if (error instanceof Error && error.message === "Operation cancelled") {
      console.log("Critical path calculation cancelled")
      return []
    }
    console.error("Error during critical path calculation:", error)
    throw error
  }
}

File: pavel-sdk/src/pavel_sdk/critical_path.py
"""Critical path calculation for dbt execution graphs."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, MutableMapping

from .data import Dataset

__all__ = ["calculate_critical_path"]


@dataclass
class _Node:
    id: str
    duration: float
    parents: List[str]
    children: List[str]
    earliest_start: float = 0.0
    earliest_finish: float = 0.0
    latest_start: float = 0.0
    latest_finish: float = 0.0


def _build_graph(dataset: Dataset) -> Dict[str, _Node]:
    nodes: Dict[str, _Node] = {}
    for record in dataset.records.to_dict("records"):
        nodes[record["model"]] = _Node(
            id=record["model"],
            duration=float(record["duration"]),
            parents=[],
            children=[],
        )

    for link in dataset.links.to_dict("records"):
        source = link.get("source")
        target = link.get("target")
        if source in nodes and target in nodes:
            nodes[source].children.append(target)
            nodes[target].parents.append(source)

    return nodes


def calculate_critical_path(dataset: Dataset) -> List[str]:
    """Return the ordered list of model ids that form the critical path."""

    if dataset.records.empty:
        return []

    graph = _build_graph(dataset)
    if not graph:
        return []

    in_degree: MutableMapping[str, int] = {node_id: len(node.parents) for node_id, node in graph.items()}
    queue = [node_id for node_id, deg in in_degree.items() if deg == 0]
    order: List[str] = []

    while queue:
        current = queue.pop(0)
        order.append(current)
        node = graph[current]
        node.earliest_finish = node.earliest_start + node.duration
        for child_id in node.children:
            child = graph[child_id]
            child.earliest_start = max(child.earliest_start, node.earliest_finish)
            in_degree[child_id] -= 1
            if in_degree[child_id] == 0:
                queue.append(child_id)

    if not order:
        return []

    project_end = max(graph[node_id].earliest_finish for node_id in graph)
    for node_id in reversed(order):
        node = graph[node_id]
        if not node.children:
            node.latest_finish = project_end
        else:
            node.latest_finish = min(graph[child].latest_start for child in node.children)
        node.latest_start = node.latest_finish - node.duration

    # Reconstruct critical path
    critical_path: List[str] = []
    current = None
    for node_id in reversed(order):
        node = graph[node_id]
        if not node.children and abs(node.earliest_finish - node.latest_finish) < 1e-6:
            current = node_id
            break

    while current:
        critical_path.insert(0, current)
        parents = graph[current].parents
        next_parent = None
        for parent in parents:
            parent_node = graph[parent]
            if abs(parent_node.earliest_finish - parent_node.latest_finish) < 1e-6 and abs(
                parent_node.earliest_finish - graph[current].earliest_start
            ) < 1e-6:
                next_parent = parent
                break
        current = next_parent

    return critical_path
